### ISSUE TITLE
firebase-tools v11.3.0 deploy fails for gen2 Cloud Functions

### ISSUE THREAD

**Issue by metadiego** [firebase/firebase-tools#5334]
### [REQUIRED] Environment info

**firebase-tools:** 11.3.0

**Platform:** macOS

### [REQUIRED] Test case

A firebase project with gen2 cloud functions and local version of firebase-tools === 11.3.0. Firebase deploy fails.

### [REQUIRED] Steps to reproduce

1. Install the latest version of Firebase CLI: ``` npm install -g firebase-tools@11.3.0 ```
2. Create a firebase project: ``` firebase init ```
3. Add a gen2 cloud function
4. Run ``` firebase deploy ```

### [REQUIRED] Expected behavior

Firebase deploy success.

### [REQUIRED] Actual behavior

Deploy fails with following error: ```Error: Failed to fetch Run service undefined```.

If we run ```firebase deploy --debug```, we get:
```
[2022-12-14T15:01:42.800Z] FirebaseError: Failed to make request: Cannot read properties of undefined (reading 'startsWith')
    at Client.request (/Users/user/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/apiv2.js:111:19)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.getService (/Users/user/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/gcp/run.js:17:26)
    at async Promise.all (index 0)
    at async loadExistingBackend (/Users/user/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/deploy/functions/backend.js:197:28)
    at async Object.existingBackend (/Users/user/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/deploy/functions/backend.js:173:9)
    at async prepare (/Users/user/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/deploy/functions/prepare.js:110:115)
    at async chain (/Users/user/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/deploy/index.js:33:9)
    at async deploy (/Users/user/.cache/firebase/tools/lib/node_modules/firebase-tools/lib/deploy/index.js:68:5)
```

### Temporary Fix:
In order to fix the error, I had to remove Firebase tools (v 11.3.0) from my system and install Firebase tools v11.18.0. Deploy succeeds.

**Comment by shin-eraser** [comment]
+1

**Comment by ojetokun** [comment]
If you've added firebase-tools to your ci/cd pipeline and it keeps failing with the same error. You just need to update the firebase-tools included in your pipeline. See [cloud.google.com/build/docs/deploying-builds/deploy-firebase](https://cloud.google.com/build/docs/deploying-builds/deploy-firebase) for more

**Comment by bmitchinson** [comment]
Need to upgrade your copy of firebase tools with `npm install -g firebase-tools` 

https://stackoverflow.com/a/74759853

**Comment by glynnallen1704** [comment]
Facing the same issue above temp fix also not working: 
DEBUG shows the following error :
FirebaseError: Failed to make request: Cannot read properties of undefined (reading 'startsWith')

**Comment by ojetokun** [comment]
> facing the same issue nothing works for me from the mentioned suggestion

Try running "$ firebase tools --version" **FROM** the project directory. What version of firebase-tools are you using?

Basically ensure that it is 11.18.0

**Comment by glynnallen1704** [comment]
@ojetokun Thank you seems like i needed to uninstall by running the script **curl -sL firebase.tools | uninstall=true bash** rather than just the npm uninstall and after doind so the correct version was installed and successfully deploying now.

**Comment by AmanDeqode** [comment]
```
Try running "$ firebase tools --version" FROM the project directory. What version of firebase-tools are you using?

Basically ensure that it is 11.18.0
```
Tried with latest version also
Actually it works at my local machine but cause issue with the github actions 
I tried this with both upgrading and degrading the version of firebase tools but didn't works for me

**Comment by AmanDeqode** [comment]
@ojetokun tried with latest version 11.18.0 breaks the code at local as well as on github actions
with 11.0.1 works fine at local breaks on github actions

**Comment by AmanDeqode** [comment]
@ojetokun it works now thanks

**Comment by zoomisen** [comment]
For me in linux (ubuntu), `firebase -V` and `firebase tools --version` showed something like v11.7 even after an update `npm -g i firebase-tools@latest`, but `npm -g list` showed version 11.18.0

I tried to remove firebase-tools with `npm -g remove firebase-tools`, but even after that the firebase command was still available and still showed v11.7

I used `which firebase` to figure out where the firebase command was located, and removed it manually `rm /usr/bin/firebase`.

And then reinstalled firebase-tools with `npm -g i firebase-tools@latest`. And after that I was able to do a deploy.

**Comment by ColtonGrady98** [comment]
> For me in linux (ubuntu), `firebase -V` and `firebase tools --version` showed something like v11.7 even after an update `npm -g i firebase-tools@latest`, but `npm -g list` showed version 11.18.0
> 
> I tried to remove firebase-tools with `npm -g remove firebase-tools`, but even after that the firebase command was still available and still showed v11.7
> 
> I used `which firebase` to figure out where the firebase command was located, and removed it manually `rm /usr/bin/firebase`.
> 
> And then reinstalled firebase-tools with `npm -g i firebase-tools@latest`. And after that I was able to do a deploy.

I'm on mac, I needed to do a similar thing. I'm using nvm, and needed to run `which firebase` twice and delete with `rm <location>` in both places before running `npm -g i firebase-tools@latest`, but that fixed my issue.

**Comment by metadiego** [comment]
Yep, I had to do a similar thing as @zoomisen and @ColtonGrady98 to get this to work. To be clear:

1. Run: ``` firebase --version ```
2. If it is 11.3.0, do whatever it takes to delete it. I had to manually remove it because ```npm uninstall -g firebase-tools``` did not work.
3. Run ``` firebase --version ``` and make sure firebase is not installed.
4. Follow the instructions here: [https://firebase.google.com/docs/cli](https://firebase.google.com/docs/cli) to install. For mac, here are 2 options:
a. ```npm install -g firebase-tools```
b. Use the install script: ```curl -sL https://firebase.tools | upgrade=true bash``` (this one worked for me).

**Comment by colerogers** [comment]
Hi everyone, sorry you folks ran into this. Just to be clear, the deployment stopped working without any changes on your side, correct? Also, what is the version of `firebase-functions` that you are all using? Thanks

**Comment by colerogers** [comment]
Also, could someone could post the entire debug log so we can get context for the failed api call

**Comment by colerogers** [comment]
After some digging, we pushed a change back in September ( #5001 ) that allowed the `firebase-tools` CLI to handle an upstream change from the Google Cloud Functions API. This was released in version `11.9.0` and it looks like the GCF team recently rolled out the change on their end. That's why you are seeing this error if you have an older `firebase-tools` version. The best course of action is to upgrade your tools version to be at least `11.9.0` but preferably the latest (which looks like most of you already have). 

Apologies for the time spent trying to debug this issue, we rarely have upstream changes that break our tooling like this.

**Comment by rendomnet** [comment]
@colerogers when we update firebase-tools then we are also forced to update firebase-admin adn firebase-functions. And this is also giving us another ERROR. i  functions: Watching ```
"C:\....\functions\dist" for Cloud Functions...
The system cannot find the path specified.

Error: An unexpected error has occurred.
```
